### PR TITLE
HPCC-32228 Add fixed_size_binary data type to Parquet Plugin 

### PR DESCRIPTION
**Pull request by ilhan2316**
## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [x] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Review by jackdelv — CHANGES_REQUESTED** [review]
@ilhan2316 I think it's close. There are just a few style issues that need to be fixed. 

There were a few cases where extra lines were added and not needed. I commented those with Extra Line.

File: plugins/parquet/parquetembed.hpp [hunk 1]
```diff
@@ -132,7 +132,8 @@ enum ParquetArrayType
     DecimalType,
     ListType,
     StructType,
-    RealType
+    RealType,
+    fixed_size_binary
```
[the remark this answers is not in the document]
Comment (jackdelv) [reply plugins/parquet/parquetembed.hpp:136]: Looking at this again it should actually be FixedSizeBinaryType like the rest of the struct members.

File: plugins/parquet/parquetembed.cpp [hunk 2]
```diff
@@ -725,11 +725,10 @@ std::shared_ptr<arrow::NestedType> ParquetWriter::makeChildRecord(const RtlField
  * @param arrowFields Output vector for pushing new nodes to.
  * @return Status of the operation
  */
-arrow::Status ParquetWriter::fieldToNode(const RtlFieldInfo *field, std::vector<std::shared_ptr<arrow::Field>> &arrowFields)
+arrow::Status ParquetWriter::fieldToNode(const RtlFieldInfo *field, std::vector<std::shared_ptr<arrow::Field>> &arrowFields)
```
Comment (jackdelv) [inline plugins/parquet/parquetembed.cpp:728]: Extra space.

File: plugins/parquet/parquetembed.cpp [hunk 3]
```diff
@@ -785,7 +784,7 @@ arrow::Status ParquetWriter::fieldToNode(const RtlFieldInfo *field, std::vector<
         failx("Datatype %i is not compatible with this plugin.", field->type->getType());
     }
 
-    return arrow::Status::OK();
+   return arrow::Status::OK();
```
Comment (jackdelv) [inline plugins/parquet/parquetembed.cpp:787]: Removed space.

File: plugins/parquet/parquetembed.cpp [hunk 4]
```diff
@@ -1178,7 +1177,9 @@ std::string_view ParquetRowBuilder::getCurrView(const RtlFieldInfo *field)
             return arrayVisitor->largeStringArr->GetView(currArrayIndex());
         case DecimalType:
             return arrayVisitor->size == 128 ? arrayVisitor->decArr->GetView(currArrayIndex()) : arrayVisitor->largeDecArr->GetView(currArrayIndex());
-        default:
+        case FixedSizeBinaryType:
+            return arrayVisitor->fixedSizeBinaryArr->GetView(currArrayIndex());
+         default:
```
Comment (jackdelv) [inline plugins/parquet/parquetembed.cpp:1182]: Extra space.

**Review by jackdelv — CHANGES_REQUESTED** [review]
@ilhan2316 It looks good. There are still a few places with trailing spaces or too few spaces. I have placed comments at those points.

I think that after these changes the documentation should be updated. Please create a JIRA for updating the Parquet documentation with this type and large_list.

File: plugins/parquet/parquetembed.cpp [hunk 5]
```diff
@@ -725,11 +725,10 @@ std::shared_ptr<arrow::NestedType> ParquetWriter::makeChildRecord(const RtlField
  * @param arrowFields Output vector for pushing new nodes to.
  * @return Status of the operation
  */
-arrow::Status ParquetWriter::fieldToNode(const RtlFieldInfo *field, std::vector<std::shared_ptr<arrow::Field>> &arrowFields)
+arrow::Status ParquetWriter::fieldToNode(const RtlFieldInfo *field, std::vector<std::shared_ptr<arrow::Field>> &arrowFields) 
 {
     StringBuffer name;
     xpathOrName(name, field);
-
```
Comment (jackdelv) [inline plugins/parquet/parquetembed.cpp:732]: This line should be added back for readability.

File: plugins/parquet/parquetembed.cpp [hunk 6]
```diff
@@ -770,10 +770,10 @@ arrow::Status ParquetWriter::fieldToNode(const RtlFieldInfo *field, std::vector<
     case type_unicode:
     case type_varunicode:
     case type_decimal:
-        arrowFields.push_back(std::make_shared<arrow::Field>(name.str(), arrow::utf8())); //TODO add decimal encoding
+        arrowFields.push_back(std::make_shared<arrow::Field>(name.str(), arrow::utf8())); // TODO: add decimal encoding
         break;
     case type_data:
-        arrowFields.push_back(std::make_shared<arrow::Field>(name.str(), arrow::large_binary()));
+        arrowFields.push_back(std::make_shared<arrow::Field>(name.str(), arrow::fixed_size_binary(field->type->length)));
```
Comment (jackdelv) [inline plugins/parquet/parquetembed.cpp:776]: I don't think we should be writing fixed_size_binary to Parquet files in all cases. I don't know the answer to this question, but what is the value of field->type->length if the DATA field is defined without the number of bytes e.g. DATA mydata vs DATA8 myfixeddata.

File: plugins/parquet/parquetembed.cpp [hunk 7]
```diff
@@ -729,7 +729,7 @@ arrow::Status ParquetWriter::fieldToNode(const RtlFieldInfo *field, std::vector<
 {
     StringBuffer name;
     xpathOrName(name, field);
-
+
```
Comment (jackdelv) [inline plugins/parquet/parquetembed.cpp:732]: Trailing spaces.

File: plugins/parquet/parquetembed.hpp [hunk 8]
```diff
@@ -270,6 +271,15 @@ class ParquetArrayVisitor : public arrow::ArrayVisitor
         size = 8;
         return arrow::Status::OK();
     }
+
```
Comment (jackdelv) [inline plugins/parquet/parquetembed.hpp:274]: Please remove the newlines before and after Visit.

**Review by jackdelv — CHANGES_REQUESTED** [review]
@ilhan2316 It's close. I just have one question that may require a little investigation.

File: plugins/parquet/parquetembed.cpp (hunk 6 above)
Comment (ilhan2316) [reply plugins/parquet/parquetembed.cpp:776]: Good question. I will look into this and get back to you.

File: plugins/parquet/parquetembed.cpp [hunk 9]
```diff
@@ -1178,8 +1190,10 @@ std::string_view ParquetRowBuilder::getCurrView(const RtlFieldInfo *field)
             return arrayVisitor->largeStringArr->GetView(currArrayIndex());
         case DecimalType:
             return arrayVisitor->size == 128 ? arrayVisitor->decArr->GetView(currArrayIndex()) : arrayVisitor->largeDecArr->GetView(currArrayIndex());
-        default:
-            failx("Unimplemented Parquet type for field with name %s.", field->name);
+         case FixedSizeBinaryType:
```
[the remark this answers is not in the document]
Comment (jackdelv) [reply plugins/parquet/parquetembed.cpp:1193]: The indent of the failx call is also off.

**Review by jackdelv — CHANGES_REQUESTED** [review]
It is very close. There are just a few style issues, but the changes look correct.

**Review by jackdelv — CHANGES_REQUESTED** [review]
One last style nitpick.

**Review by jackdelv — APPROVED** [review]
@ilhan2316 The changes look good.

**Review by dcamper — APPROVED** [review]
Looks good.

**Comment by dcamper** [comment]
Offline commented noted here for posterity and note-keeping: Investigation into using <vector>.emplace_back() instead of <vector>.push_back() may be a good idea.

**Comment by jackdelv** [comment]
@ilhan2316 Please squash to a single commit.

**Comment by ilhan2316** [comment]
@ghalliday This is ready to be merged. Please take a look.

**Comment by ghalliday** [comment]
@ilhan2316 please can you squash the commits

**Comment by ilhan2316** [comment]
@ghalliday I have squashed.

**Comment by ghalliday** [comment]
@jackdelv please can you review/approve

**Review by jackdelv — APPROVED** [review]
@ghalliday The changes are correct.

**Pull request merged**